### PR TITLE
Issue 4874: Update hadoopVersion

### LINKED ISSUE #4874: Fix HDFS security vulnerabilities
**Problem description**
Pravega uses HDFS for tier 2 and it has a set of security vulnerabilities. These include `CVE-2016-6811, CVE-2018-8009, CVE-2018-8029, CVE-2018-11766`. These come from the Apache Hadoop hdfs library along with its transitive dependencies. The versions of these libraries needs to be updated to overcome the security vulnerabilities. The updates to the library requires changes in Pravega source code to remove the methods, exceptions and properties used from the library that are no longer available and to use the replacements for these and other deprecated syntax.

**Problem location**
The version for the Apache hadoop hdfs library is specified in the gradle.properties. Its usage spans HDFS tests and exception helpers. These changes expand to others parts of the code where they are used. 

**Suggestions for an improvement**
 The version 2.9.2 of apache hadoop hdfs fixes the above CVEs but there are a few more CVEs that have been addressed in recent versions. These new versions have increased the breaks in compatibility with our existing code. The updates can therefore be progressive increasing from 2.9.2 to more recent versions. The changes to the hdfs library and its transitive dependencies can be made at once.

### PR DESCRIPTION
**Pull request by ravibeta**
**Change log description**  
This touches only one file gradle.properties to update the version of pravega-hdfs.
The hadoopVersion is updated to 2.9.2 as described in the linked issue. However, the guava version had to be lowered slightly. This is not a mistake. All revisions upwards of 2.9.2 up to the latest have had this breaking change which is described by the issues [HDFS-752](https://issues.apache.org/jira/browse/HDFS-752) and [HDFS-7477](https://issues.apache.org/jira/browse/HDFS-7477) and they don't agree with guava. The links for these issues are mentioned here so that the patch associated with these issues can be viewed. They are the only break-in changes we have. These changes are applicable to all recent versions but 2.9.2 as described in the attached issue fixes all the security vulnerabilities cited in the GitHub issue [4874](https://github.com/pravega/pravega/issues/4874) linked to this request.  

**Purpose of the change**  
The purpose of this change is to remove the security vulnerabilities associated with the hadoop version. The Hadoop Version poses a bigger security threat than others to Pravega.

Fixes #4874.

**What the code does**  
This code just changes versions in gradle.properties

**How to verify it**  
This change was verified by travis build.

The exception faced with upgrading Guava is
```
io.pravega.storage.hdfs.HDFSStorageTest > testDelete FAILED
    java.lang.NoSuchMethodError: com.google.common.util.concurrent.Futures.addCallback(Lcom/google/common/util/concurrent/ListenableFuture;Lcom/google/common/util/concurrent/FutureCallback;)V
        at org.apache.hadoop.hdfs.server.datanode.checker.ThrottledAsyncChecker.addResultCachingCallback(ThrottledAsyncChecker.java:167)
        at org.apache.hadoop.hdfs.server.datanode.checker.ThrottledAsyncChecker.schedule(ThrottledAsyncChecker.java:156)
        at org.apache.hadoop.hdfs.server.datanode.checker.StorageLocationChecker.check(StorageLocationChecker.java:164)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.makeInstance(DataNode.java:2762)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.instantiateDataNode(DataNode.java:2677)
        at org.apache.hadoop.hdfs.MiniDFSCluster.startDataNodes(MiniDFSCluster.java:1663)
        at org.apache.hadoop.hdfs.MiniDFSCluster.initMiniDFSCluster(MiniDFSCluster.java:905)
        at org.apache.hadoop.hdfs.MiniDFSCluster.<init>(MiniDFSCluster.java:517)
        at org.apache.hadoop.hdfs.MiniDFSCluster$Builder.build(MiniDFSCluster.java:476)
        at io.pravega.storage.hdfs.HDFSClusterHelpers.createMiniDFSCluster(HDFSClusterHelpers.java:32)
        at io.pravega.storage.hdfs.HDFSStorageTest.setUp(HDFSStorageTest.java:68)
``` 
This method was removed from this library as seen by [others](https://issues.jenkins-ci.org/browse/JENKINS-56252) as well. This [issue](https://issues.apache.org/jira/browse/BEAM-3156) is tracked on the originator side as well with this [jira](https://issues.apache.org/jira/browse/HADOOP-16765). Notice the method takes two parameters instead of the [three](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/util/concurrent/FutureCallback.html) in recent versions. The usual resolution of using exclusions, diff'ing dependency trees and ordering dependency load such as this [one](https://stackoverflow.com/questions/8257545/nosuchmethoderror-exception-when-using-com-google-common-base-splitter/23253320#23253320) do not work.

Testing:
```
docker stack services pravega
ID                  NAME                   MODE                REPLICAS            IMAGE                                                                          PORTS
127hedtw2lzp        pravega_hdfs           replicated          1/1                 pravega/hdfs:2.7.7                                                        
182fk0erog8d        pravega_segmentstore   replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:12345->12345/tcp
lcdxxqpdx3h9        pravega_zookeeper      replicated          1/1                 zookeeper:3.5.4-beta                                                      
m9xy1i32zerv        pravega_bookkeeper     replicated          3/3                 devops-repo.isus.emc.com:8116/rajamr2/bookkeeper:0.8.0-2577.5d687b9-SNAPSHOT
y8vrw86ysu7y        pravega_controller     replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:9090->9090/tcp,*:10080->10080/tcp

```
Note that the exception described above is being addressed with different jars where the newer one is different from the older one) and the guava blogs also calls out this issue:
https://mvnrepository.com/artifact/com.google.guava/listenablefuture
[Downgrading](https://blog.gradle.org/version-downgrade) is the only option left at this time.
[GuavaVersion29versus25.txt](https://github.com/pravega/pravega/files/4815622/GuavaVersion29versus25.txt)

Unit-testing:
```
404 tests completed, 1 failed
StreamSegmentContainerTests. testWriteFenceOut
org.junit.runners.model.TestTimedOutException: test timed out after 100000 milliseconds
#  rerunning
io.pravega.segmentstore.server.containers.StreamSegmentContainerTests > testWriteFenceOut PASSED
```

Testing with 3.1.1:
```
docker stack services pravega
ID                  NAME                   MODE                REPLICAS            IMAGE                                                                          PORTS
c0ivwwzvonrs        pravega_zookeeper      replicated          1/1                 zookeeper:3.5.4-beta                                                      
dgolahviacf7        pravega_hdfs           replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/hdfs:3.1.1                                                       
n5114asbhf90        pravega_controller     replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:9090->9090/tcp,*:10080->10080/tcp
o5yz5hmaan8d        pravega_bookkeeper     replicated          3/3                 devops-repo.isus.emc.com:8116/rajamr2/bookkeeper:0.8.0-2577.5d687b9-SNAPSHOT
qo9ushm24t2f        pravega_segmentstore   replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:12345->12345/tcp

```

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -31,8 +31,8 @@ gradleLombokPluginVersion=3.2.0
 gradleMkdocsPluginVersion=1.0.0
 gradleSshPluginVersion=2.9.0
 grpcVersion=1.17.1
-guavaVersion=28.2-jre
-hadoopVersion=2.8.5
+guavaVersion=25.0-jre
```
Comment (andreipaduroiu) [inline gradle.properties:34]: We should not be going back to an older version of guava.

Comment (ravibeta) [reply gradle.properties:34]: Yes Andrei, there was no other choice. All  newer versions of guava do not agree with hadoopVersion 2.9.2 or later.

**Comment by eolivelli** [comment]
Is HDFS client 3.2.x compatible with HDFS 2 servers?

If this is true that I suggest you to upgrade to the latest version.
We will still be using HDFS 2 in docker images so we can guarantee the compatibility with existing Pravega clusters with HDFS

**Comment by ravibeta** [comment]
Tested with docker swarm:
```
docker stack services pravega
ID                  NAME                   MODE                REPLICAS            IMAGE                                                                          PORTS
127hedtw2lzp        pravega_hdfs           replicated          1/1                 pravega/hdfs:2.7.7                                                        
182fk0erog8d        pravega_segmentstore   replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:12345->12345/tcp
lcdxxqpdx3h9        pravega_zookeeper      replicated          1/1                 zookeeper:3.5.4-beta                                                      
m9xy1i32zerv        pravega_bookkeeper     replicated          3/3                 devops-repo.isus.emc.com:8116/rajamr2/bookkeeper:0.8.0-2577.5d687b9-SNAPSHOT
y8vrw86ysu7y        pravega_controller     replicated          1/1                 devops-repo.isus.emc.com:8116/rajamr2/pravega:0.8.0-2577.5d687b9-SNAPSHOT      *:9090->9090/tcp,*:10080->10080/tcp

```

**Comment by ravibeta** [comment]
@andreipaduroiu I have made this fix test only. Thanks, Ravi

**Comment by eolivelli** [comment]
I am not sure this is a good way to good.
Having different JARs on classpath during tests is dangerous, your test will say that everything works but in production it won't run.

Is guava downgrade needed only to run an embedded HDFS system ? if it is used by the client than we must have the same version in production on the segmentstore

**Comment by ravibeta** [comment]
> I am not sure this is a good way to good.
> Having different JARs on classpath during tests is dangerous, your test will say that everything works but in production it won't run.
> 
> Is guava downgrade needed only to run an embedded HDFS system ? if it is used by the client than we must have the same version in production on the segmentstore

File: build.gradle [hunk 2]
```diff
@@ -402,6 +422,16 @@ project('segmentstore:server:host') {
         into "conf"
     }
 
+
+    configurations.testRuntimeClasspath {
+        resolutionStrategy {
+                force "com.google.inject.extensions:guice-assistedinject:3.0"
+                force "com.google.inject.extensions:guice-servlet:3.0"
+                force "com.google.inject:guice:3.0"
```
Comment (RaulGracia) [inline build.gradle:430]: Should this be `guiceVersion`?

File: build.gradle [hunk 3]
```diff
@@ -683,6 +713,15 @@ project('standalone') {
         }
     }
 
+    configurations.testRuntimeClasspath {
+        resolutionStrategy {
+                force "com.google.inject.extensions:guice-assistedinject:" + guiceVersion
+                force "com.google.inject.extensions:guice-servlet:" + guiceVersion
+                force "com.google.inject:guice:" + guiceVersion
+                force "com.google.guava:guava:25.0-jre"
```
Comment (RaulGracia) [inline build.gradle:721]: Having hardcoded versions here is not a good idea. Normally we try to parameterize all these versions.

**Comment by ravibeta** [comment]
@andreipaduroiu  Igor suggested that i put it down here that 29 to 27 downgrade from guava was okayed but the change from 27 to 25 is negligible.

**Comment by medvedevigorek** [comment]
@ravibeta what about proposal to use a separate class loader for `MiniDFSCluster` in `HDFSIntegrationTest`? Could it be a working solution that will allow us to keep latest `Guava` library in the production code and use previous version just with `MiniDFSCluster` instance in the integration test?

**Comment by padma81** [comment]
The Hadoop libraries are being upgraded to 3.1.1 version. But there is a CVE against the 3.1.1 version - CVE-2018-11768. More details are here: https://hadoop.apache.org/cve_list.html.

**Please upgrade the Hadoop libraries to 3.1.2 version.**

**Comment by eolivelli** [comment]
Is there any ticket on HDFS for the upgrade of Guava?
We could create one, request for impl or even contribute that change.

I suggest not to play with the classloader with HDFS, it is a quite complex system, full  of jvm hacks.
We would have to guarantee that in the test our client is using the correct version of Guava and the server uses it's own version.

Maybe another option can be to create a new module that encapsulates HDFS mini cluster and relocated Guava classes.
I am not a Gradle expert but with Maven would've easy using the Maven shade plugin

**Comment by ravibeta** [comment]
Thanks for letting me know, I will do the following:
1) check byte-by-byte parity in stream persistence between the usages of guava 25 and guava 28
2) update hadoopVersion from 3.1.1 to 3.1.2

File: gradle.properties (hunk 1 above)
Comment (ravisharda) [reply gradle.properties:35]: @ravibeta as per https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-hdfs/3.1.2 > `Compile Dependencies` section, v3.1.2 depends on v`29.0-jre` of Guava. So, with the HDFS upgrade to this `3.1.2` you might be able to upgrade Guava now.

Comment (ravibeta) [reply gradle.properties:35]: Yes Ravi

**Comment by ravisharda** [comment]
@ravibeta 

Please avoid using the option to **re-request the review**, as it causes the block (for requested changes) on the PR to be removed. Plus, it also removes the indicator that changes were requested on the PR. 

Most reviewers listen for comments on the PRs they are reviewing, so you don't need to re-request for reviews and responding to review comments is sufficient, although in cases of delays you might want to follow-up over Slack too. 

On a different note, the code coverage target is also not met (it's 33.33% against the required 84.39%). Please address that.

**Comment by ravisharda** [comment]
@ravibeta curious, how have you testing the changes being made in this PR. 

If you haven't tried running an HDFS-based Pravega cluster, I'd suggest using Docker Compose as described [here](https://github.com/pravega/pravega/tree/master/docker/compose) and using YAML file https://github.com/pravega/pravega/blob/master/docker/compose/docker-compose.yml, which in turn deploys an HDFS cluster.

**Comment by ravibeta** [comment]
Yes Ravi, I did as you said.
[pravega_segmentstore.0.8.0-2621.3764976c5-github-issue-4874-fix.log](https://github.com/pravega/pravega/files/5014273/pravega_segmentstore.0.8.0-2621.3764976c5-github-issue-4874-fix.log)
[pravega_controller.0.8.0-2621.3764976c5-github-issue-4874-fix.log](https://github.com/pravega/pravega/files/5014274/pravega_controller.0.8.0-2621.3764976c5-github-issue-4874-fix.log)
[segment_store_pod_description.txt](https://github.com/pravega/pravega/files/5014287/segment_store_pod_description.txt)

**Comment by medvedevigorek** [comment]
There is a parallel effort going on to extract Pravega LTS bindings to a separate repository and updating hadoop libraries and other 3rd party dependencies related to specific bindings would be easier to do there.

**Comment by ravibeta** [comment]
Closed in favor of https://github.com/pravega/pravega/pull/4908/

**Pull request closed without merging**